contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+    ) -> Result<(), ContractError> {
+        if coin.amount < self.get_balance(storage) {
+            if let Some(querier) = querier {
+                let msg = MixnetExecuteMsg::DelegateToMixnodeOnBehalf {
+                    mix_identity: mix_identity.clone(),
+                    delegate_addr: self.address.clone(),
+                    coin: coin.clone(),
+                };
+                querier.query_wasm_smart(DEFAULT_MIXNET_CONTRACT_ADDRESS, &msg)?;
+            }
+            self.track_delegation(env.block.time, mix_identity, coin, storage)?;
+            Ok(())
+        } else {
+            return Err(ContractError::InsufficientBalance(
+                self.address.as_str().to_string(),
+                self.get_balance(storage).u128(),
+            ));
+        }
+    }
+
+    fn try_undelegate_from_mixnode(
+        &self,
+        mix_identity: IdentityKey,
+        storage: &mut dyn Storage,
+        querier: Option<QuerierWrapper>,
+    ) -> Result<(), ContractError> {
+        if let Some(querier) = querier {
+            let msg = MixnetExecuteMsg::UnDelegateFromMixnodeOnBehalf {
+                mix_identity: mix_identity.clone(),
+                delegate_addr: self.address.clone(),
+            };
+            querier.query_wasm_smart(DEFAULT_MIXNET_CONTRACT_ADDRESS, &msg)?;
+        }
+        self.track_undelegation(mix_identity, storage)?;
+        Ok(())
+    }
+
+    fn track_delegation(
+        &self,
+        block_time: Timestamp,
+        mix_identity: IdentityKey,
+        delegation_amount: Coin,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        let mut delegations =
+            if let Some(delegations) = get_account_delegations(storage, &self.address) {
+                delegations
+            } else {
+                Vec::new()
+            };
+        delegations.push(DelegationData {
+            mix_identity,
+            amount: delegation_amount.amount,
+            block_time,
+        });
+        set_account_delegations(storage, &self.address, delegations)?;
+        Ok(())
+    }
+
+    fn track_undelegation(
+        &self,
+        mix_identity: IdentityKey,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        // This has to exist in storage at this point.
+        let delegations = get_account_delegations(storage, &self.address)
+            .unwrap()
+            .into_iter()
+            .filter(|d| d.mix_identity != mix_identity)
+            .collect();
+        // Since we're always removing the entire delegation we can just drop the key
+
+        Ok(set_account_delegations(
+            storage,
+            &self.address,
+            delegations,
+        )?)
+    }
+}
+
+impl VestingAccount for PeriodicVestingAccount {
+    fn locked_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        // Returns 0 in case of underflow.
```

Comment:
Under what conditions could it possible underflow? Shouldn't we be accounting for every single token?

Reply:
Thats just a comment to draw attention to`saturating_sub`, in practice `delegated_vesting` can never be greater then `total_vesting`, I'll add a test around this